### PR TITLE
[HOTFIX][DEV-5600] Optimize CSV data processing with vectorized use of Pandas

### PR DESCRIPTION
**Pull request by kwhickey**
**High level description:**
Addressing bottlenecks in ingest and processing of (very large) CSV files using Pandas data processing framework. Swapping use of code doing element-wise function calls (on each cell) with vectorized calls across column Series to reduce processing time.

**Technical details:**
Pandas is meant to be used to best effect when vectorizing functions across Series. Pandas DataFrames can be thought of as a matrix of Series, which are just numpy arrays, as rows/columns. As numpy arrays, you can apply a function in a vectored way to have it take effect on the whole array at once, rather than iterating cells in that Series/array.
We seemed to be following a pattern of applying validation functions element-wise. Handing a function to pandas that we then apply row by row, cell by cell. Things that could be validated one whole column at a time.

Main validations vectorized, in new functions, to biggest effect in this PR:
1. `clean_numbers`
2. `clean_cols`
3. `check_type`
4. `check_length` 

Additional unit tests and integration tests were added to verify corner cases with `None`, `numpy.NaN`, and mixed-type DataFrames.

**Link to JIRA Ticket:**
[DEV-5600](https://federal-spending-transparency.atlassian.net/browse/DEV-1234)

The following are ALL required for the PR to be merged:
- [x] Backend review completed
- [NA] Merged concurrently with [Frontend#1234](https://github.com/fedspendingtransparency/data-act-broker-web-app/pull/1234)
- [x] Unit & integration tests updated with relevant test cases
- [NA] Frontend impact assessment completed
- [NA] Documentation Updated

### CODE REVIEW

File: dataactvalidator/validation_handlers/validationManager.py [hunk 1]
```diff
@@ -11,9 +11,10 @@
```
Comment (kwhickey) [inline dataactvalidator/validation_handlers/validationManager.py:11]: Best to look at this file with the diff settings above set to ignore whitespace (changed python indent)

File: dataactbroker/helpers/validation_helper.py [hunk 2]
```diff
@@ -80,6 +81,41 @@ def clean_col(value, clean_quotes=True):
...
+            frame: pd.DataFrame to clean
+            convert_to_str: whether to apply .astype(str) on all Series in the DataFrame
+                - If the DataFrame was created or read in from CSV with dtype=str already, this may be unnecessary
+                  and will save on performance if avoided
+                - If left as False and there are mixed types, such as int or float, they will be set to None rather
+                  than converted to their str form
+            clean_quotes: whether to clean extra quotes or not
+
+        Returns:
+            The cleaned DataFrame, with a side-effect of an
```
Comment (tony-sappe) [inline dataactbroker/helpers/validation_helper.py:99]: "The cleaned DataFrame, with a side-effect of an"

What.... What is the side-effect!?!??!

Comment (kwhickey) [reply dataactbroker/helpers/validation_helper.py]: ha... good catch. Updated docstrings.

File: dataactvalidator/validation_handlers/validationManager.py [hunk 3]
```diff
@@ -24,7 +27,8 @@
     populate_job_error_info, get_action_dates
 )
 
-from dataactcore.models.domainModels import matching_cars_subquery, Office, concat_tas_dict, concat_display_tas_dict
+from dataactcore.models.domainModels import concat_tas_dict_vectorized
+from dataactcore.models.domainModels import Office, concat_display_tas_dict
```
Comment (dpb-bah) [inline dataactvalidator/validation_handlers/validationManager.py:31]: merge?

File: dataactbroker/helpers/validation_helper.py [hunk 4]
```diff
@@ -126,6 +209,23 @@ def derive_unique_id(row, is_fabs):
...
+            is_fabs: a boolean indicating if the submission is a FABS submission or not
+
+        Returns:
+            A Series with properly formatted unique ID all rows depending on if it's a FABS or DABS submission
+    """
+    if not is_fabs:
+        return 'TAS: ' + frame['display_tas'][frame['display_tas'].notnull()].astype(str)
+    else:
+        return 'AssistanceTransactionUniqueKey: ' + \
+               frame['afa_generated_unique'][frame['afa_generated_unique'].notnull()].astype(str)
```
[the remark this answers is not in the document]
Comment (kwhickey) [reply dataactbroker/helpers/validation_helper.py:226]: From what I could tell briefly reading the docs on Pandas, string concatenation of a column (a whole Series) and not just a value, is done like this using `+`. I couldn't find explicit docs on comparing to .format, but I was afraid it would "un"-vectorize it if causing each element to be passed to a string-formatting function. 

My guess is that pandas has overloaded the `+` operator on strings to do some kind of "add this to all strings in the Series" in one vectorized operation.

To be safe, and get the most out of performance I think we should stick to this kind of string concat until we can find docs that say how using formatting / interpolation is ok.

File: dataactvalidator/validation_handlers/validationManager.py (hunk 3 above)
Comment (kwhickey) [reply dataactvalidator/validation_handlers/validationManager.py:31]: ah, good point. Probably a result of git merges. Not sure why flake didn't flake-out on that. Fixing.

File: dataactbroker/helpers/validation_helper.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dpb-bah) [reply dataactbroker/helpers/validation_helper.py:226]: Works for me 👍

[the remark this answers is not in the document]
Comment (tony-sappe) [reply dataactbroker/helpers/validation_helper.py:226]: I haven't heard about a security concern, just a (very tiny) performance hit
